### PR TITLE
Feature/issue states

### PR DESCRIPTION
**Pull request by abubakarazeem**
Issue States Controller
Added CRUD Operations using ajax
Added test cases for controller, model, abilities and routes
Only admins have the ability to access

### CODE REVIEW

**Comment by abubakarazeem** [comment]
Review: Cycle-0

File: app/assets/javascripts/issue_states.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  });
+
+
+  function clearFlashMessages() {
+    issueStateFlash.html('');
+    issueStatesErrors.html('');
+  }
+
+  function hideIssueSearch() {
+    issueNameSearch.attr("hidden",true);
```
Comment (abubakarazeem) [inline app/assets/javascripts/issue_states.js]: spacing

File: app/assets/javascripts/issue_states.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  function resetAttrIssueStateForm(){
+    issueStateForm.attr("method", issueStateFormMethod);
```
Comment (abubakarazeem) [inline app/assets/javascripts/issue_states.js]: do this using ajax

File: app/assets/javascripts/issue_states.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+});
+
+function resetIssueIssueStateCategory(){
+  issueIssueStateCategory.find('label:eq(1)').removeClass('active');
+  issueIssueStateCategory.find('label:eq(0)').addClass('active');
+}
+
+function showIssueSearch() {
+  issueNameSearch.attr("hidden",false);
```
Comment (abubakarazeem) [inline app/assets/javascripts/issue_states.js]: hide()

File: app/assets/javascripts/issue_states.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  issueIssueStateCategory.find('label:eq(1)').removeClass('active');
+  issueIssueStateCategory.find('label:eq(0)').addClass('active');
+}
+
+function showIssueSearch() {
+  issueNameSearch.attr("hidden",false);  
+}
+
+function clearIssueStateForm() {
+  issueStateForm.trigger("reset");
```
Comment (abubakarazeem) [inline app/assets/javascripts/issue_states.js]: ajax call to new

File: app/controllers/issue_states_controller.rb [hunk 5]
```diff
@@ -0,0 +1,66 @@
+class IssueStatesController < ApplicationController
+  autocomplete :issue, :title, display_value: :title_with_project_id, extra_data: [:project_id]
+  load_and_authorize_resource
+
+  def index
+    @issue_states = @issue_states.issue_specific_states(params[:issue_id]) unless params[:issue_id].nil?
```
Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: try to do this using cancan ability

Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: use if with present

File: app/controllers/issue_states_controller.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    # required for new_issue_state_modal
```
Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: remove

File: app/controllers/issue_states_controller.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    @issue_state = IssueState.new
```
Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: use new action

File: app/controllers/issue_states_controller.rb [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  def show
+    @issue = @issue_state.issue if @issue_state.issue_id
```
Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: remove if condition

File: app/controllers/issue_states_controller.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  end
+
+  def edit
+    respond_to do |format|
+      format.js
+    end
+  end
+
+  def update
+    if @issue_state.safe_update?(issue_state_params)
```
Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: can_update?

Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: check_associations_validation

File: app/controllers/issue_states_controller.rb [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    else
+      flash.now[:danger] = t('.not_updated')
+    end
+    respond_to do |format|
+      format.js
+    end
+  end
+
+  def destroy
+    if @issue_state.safe_destroy?
```
Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: use validations.

File: app/controllers/issue_states_controller.rb [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      flash.now[:danger] = t('.not_deleted')
+    end
+    respond_to do |format|
+      format.js
+    end
+  end
+
+  def issue_state_params
+    result = params.require(:issue_state).permit(:name, :issue_id)
+    result[:issue_id] = params[:issue_id] if params[:issue_id].present? && params[:category] == 'issue'
```
Comment (abubakarazeem) [inline app/controllers/issue_states_controller.rb]: verify if we can do this through new method

File: app/helpers/issue_states_helper.rb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+module IssueStatesHelper
```
Comment (abubakarazeem) [inline app/helpers/issue_states_helper.rb]: remove this

File: app/models/issue_state.rb [hunk 13]
```diff
@@ -1,7 +1,44 @@
...
+  validates :company_id, presence: true
+
   has_many :issues
   belongs_to :company
   belongs_to :issue
 
+  scope :issue_specific_states, ->(issue_id) { where(issue_id: [issue_id, nil]) }
   scope :issue_states_for_projects, ->(project) { project.issues.map(&:issue_state) }
+
+  def safe_update?(params)
```
Comment (abubakarazeem) [inline app/models/issue_state.rb]: change name and use validations

File: app/views/issue_states/_issue_state.html.erb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+<tr id="issue_state_<%= issue_state.id %>" class="<%= issue_state.issue_id.nil? ? 'global' : 'issue-'+issue_state.issue_id.to_s %>">
+  <td><%= issue_state.name %></td>
+  <td> <% if issue_state.issue_id.nil? %>
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state.html.erb]: indentation

File: app/views/issue_states/_issue_state.html.erb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+    <i class="fas fa-check-circle"></i>
+    <% else %>
+    <i class="fas fa-times-circle"></i>
+    <% end %>
+  </td>
+  <td><%= issue_state.issue_id.nil? ? 'NA' : issue_state.issue_id.to_s %></td>
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state.html.erb]: helper method

File: app/views/issue_states/_issue_state_form.html.erb [hunk 16]
```diff
@@ -0,0 +1,48 @@
+<%= form_for(@issue_state, url: params[:issue_id].nil? ? issue_states_path : issue_issue_states_path(params[:issue_id]), html: {id: 'issue_state_form'}, remote: true) do |f| %>
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: change this using new action

Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: helper method

File: app/views/issue_states/_issue_state_form.html.erb [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+<% if params[:issue_id].nil? %>
+<div class="input-group mb-4">
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: indentation

File: app/views/issue_states/_issue_state_form.html.erb [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+  <div class="input-group-prepend">
+    <label class="input-group-text" for="issue_state_category">Category</label>
+  </div>
+  <%= select_tag("issue_state_category", raw("<option value='global' selected>Global</option><option value='issue'>Issue Specific</option>"), class: "custom-select") %>
+</div>
+
+<% else %>
+<div class="row">
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: indentation

File: app/views/issue_states/_issue_state_form.html.erb [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+  <div class="col-lg-12">
+    <div class="btn-group btn-group-toggle float-right" id="issue_issue_state_category" data-toggle="buttons">
+      <label class="btn btn-info btn-sm active">
+        <!-- TODO: change category to unique name -->
+        <input type="radio" name="category" value="issue" checked> Issue Specific
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: change category name to unique name

File: app/views/issue_states/_issue_state_form.html.erb [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+  <div class="input-group-prepend">
+    <label class="input-group-text" for="issue_state_category">Category</label>
+  </div>
+  <%= select_tag("issue_state_category", raw("<option value='global' selected>Global</option><option value='issue'>Issue Specific</option>"), class: "custom-select") %>
+</div>
+
+<% else %>
+<div class="row">
+  <div class="col-lg-12">
+    <div class="btn-group btn-group-toggle float-right" id="issue_issue_state_category" data-toggle="buttons">
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: change id name

File: app/views/issue_states/_issue_state_form.html.erb [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+        <input type="radio" name="category" value="global"> Global
+      </label>
+    </div>
+  </div>
+</div>
+<% end %>
+
+<div class="form-group">
+  <%= f.label :name, 'Issue State Name:'%>
+  <%= f.text_field :name, class: 'form-control', placeholder: 'Enter Name', required: true%>
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: spacing

File: app/views/issue_states/_issue_state_form.html.erb [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+</div>
+<% if params[:issue_id].nil? %>
+<div id="issue_name_search" hidden>
+  <label for="issue_div" class="mt-2">Choose Issue:</label>
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: change id name

File: app/views/issue_states/_issue_state_form.html.erb [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+  <div id="issue_div">
+    <div class="input-group mb-4">
+      <div class="input-group-prepend">
+        <label class="input-group-text bg-dark text-white" for="issue_state_issue_name">Issue</label>
+      </div>
+      <%= autocomplete_field_tag 'issue_state_issue_name', '', autocomplete_issue_title_issue_states_path, :id_element => '#issue_state_issue_id', class: 'form-control', placeholder: 'Search' %>
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: new hash syntax

File: app/views/issue_states/_issue_state_form.html.erb [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+    </div>
+    <%= f.number_field :issue_id, class: 'form-control', readonly: true, hidden: true%>
```
Comment (abubakarazeem) [inline app/views/issue_states/_issue_state_form.html.erb]: space

File: app/views/issue_states/_show_issue_state.html.erb [hunk 25]
```diff
@@ -0,0 +1,14 @@
+<div class="container">
+  <div class="jumbotron">
+    <h1 class="text-center">
+      <%= @issue_state.name %><span class="badge badge-pill badge-dark" style="font-size: 0.3em;"><%= @issue.nil? ? 'global' : link_to('issue', '#', class: 'text-white') %></span>
+    </h1>      
+    <p class="text-center"><span class="badge badge-pill badge-danger"><%= @total_issues %></span><%= " issue".pluralize(@total_issues) + " using this state" %>
+    </p>
+  </div>
+
+  <% unless @issue.nil? %>
```
Comment (abubakarazeem) [inline app/views/issue_states/_show_issue_state.html.erb]: if with present

File: app/views/issue_states/_show_issue_state.html.erb [hunk 26]
```diff
@@ -0,0 +1,14 @@
+<div class="container">
+  <div class="jumbotron">
+    <h1 class="text-center">
+      <%= @issue_state.name %><span class="badge badge-pill badge-dark" style="font-size: 0.3em;"><%= @issue.nil? ? 'global' : link_to('issue', '#', class: 'text-white') %></span>
```
Comment (abubakarazeem) [inline app/views/issue_states/_show_issue_state.html.erb]: line break, same for line 6

File: app/views/issue_states/_show_issue_state.html.erb (hunk 25 above)
Comment (abubakarazeem) [inline app/views/issue_states/_show_issue_state.html.erb]: indentation

File: app/views/issue_states/_show_issue_state_modal.html.erb [hunk 27]
```diff
@@ -0,0 +1,17 @@
+<div class="modal fade" id="show_issue_state_modal" tabindex="-1" role="dialog" aria-hidden="true">
+  <div class="modal-dialog" role="document">
+    <div class="modal-content">
+      <div class="modal-header">
+        <h5 class="modal-title" id="exampleModalLabel">Issue State</h5>
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+          <span aria-hidden="true">&times;</span>
+        </button>
+      </div>
+      <div class="modal-body">
```
Comment (abubakarazeem) [inline app/views/issue_states/_show_issue_state_modal.html.erb]: use a div to render something

File: app/views/issue_states/create.js.erb [hunk 28]
```diff
@@ -0,0 +1,9 @@
+issueStateFlash.html('<%= j render partial: "layouts/notice" %>');
+<% unless @issue_state.errors.any? %>
```
Comment (abubakarazeem) [inline app/views/issue_states/create.js.erb]: use if

File: app/views/issue_states/create.js.erb [hunk 29]
```diff
@@ -0,0 +1,9 @@
...
+  clearIssueStateForm();
+  issueStatesTableBody.append('<%= j render @issue_state %>');
+  tr = $("#issue_state_<%= @issue_state.id %>");
+  issueStatesTable.row.add(tr).draw();
+<% else %>
+  issueStatesErrors.html('<%= j render partial: "layouts/errors", locals: {model: @issue_state} %>');
```
Comment (abubakarazeem) [inline app/views/issue_states/create.js.erb]: space

File: app/views/issue_states/destroy.js.erb [hunk 30]
```diff
@@ -0,0 +1,6 @@
+$("#flash_messages").html('<%= j render partial: "layouts/notice" %>');
+$("#error_messages").html('<%= j render partial: "layouts/errors", locals: {model: @issue_state} %>');
+
+<% unless @issue_state.errors.any? %>
```
Comment (abubakarazeem) [inline app/views/issue_states/destroy.js.erb]: if,
space on next line

File: app/views/issue_states/edit.js.erb [hunk 31]
```diff
@@ -0,0 +1,9 @@
+<% if params[:issue_id].nil? && @issue_state.issue_id.present? %>
```
Comment (abubakarazeem) [inline app/views/issue_states/edit.js.erb]: use new action and edit action to render form

File: app/views/issue_states/edit.js.erb [hunk 32]
```diff
@@ -0,0 +1,9 @@
...
+issueStateCategory.val('issue');
```
Comment (abubakarazeem) [inline app/views/issue_states/edit.js.erb]: use partials to render different forms

File: app/views/issue_states/edit.js.erb [hunk 33]
```diff
@@ -0,0 +1,9 @@
...
+showIssueSearch();
+issueStateIssueId.val("<%= @issue_state.issue_id %>");
+<% end %>
+$('#issue_state_name').val("<%= @issue_state.name %>");
+$('#issue_state_modal').modal('show'); 
+issueStateForm.attr("method","patch");
```
Comment (abubakarazeem) [inline app/views/issue_states/edit.js.erb]: space

File: app/views/issue_states/index.html.erb [hunk 34]
```diff
@@ -0,0 +1,61 @@
+<div class="text-center my-4">
+  <h2>Issue States</h2>
+</div>
+<button type="button" class="btn btn-primary mb-4 mr-2" data-toggle="modal" data-target="#issue_state_modal">
+  <i class="fas fa-plus mr-2"></i>New Issue State</button>
+  <% if params[:issue_id].present? %>
+  <a class="btn btn-warning mb-4" href="<%= issue_states_path %>">
```
Comment (abubakarazeem) [inline app/views/issue_states/index.html.erb]: indentation

File: app/views/issue_states/index.html.erb [hunk 35]
```diff
@@ -0,0 +1,61 @@
...
+    <i class="fas fa-cog mr-2"></i>Manage All Issue States</a>
+    <% end %>
+
+
+    <!-- Issue State Modals -->
+    <%= render 'issue_states/new_issue_state_modal' %>
```
Comment (abubakarazeem) [inline app/views/issue_states/index.html.erb]: use div and render using ajax call

File: app/views/issue_states/index.html.erb [hunk 36]
```diff
@@ -0,0 +1,61 @@
...
+    <%= render 'issue_states/show_issue_state_modal' %>
```
Comment (abubakarazeem) [inline app/views/issue_states/index.html.erb]: indentation

File: app/views/issue_states/update.js.erb [hunk 37]
```diff
@@ -0,0 +1,40 @@
+issueStateFlash.html('<%= j render partial: "layouts/notice" %>');
+<% unless @issue_state.errors.any? %>
```
Comment (abubakarazeem) [inline app/views/issue_states/update.js.erb]: use if

File: app/views/issue_states/update.js.erb [hunk 38]
```diff
@@ -0,0 +1,40 @@
...
+
+function updatedIssueStateButtons() {
+  return "<%= j link_to("Delete", issue_state_path(@issue_state), method: "delete", remote: true, data: { confirm: "Are you sure to delete this issue state?"}, class: "btn btn-danger btn-sm text-white mr-2 mb-1 mb-md-0 fixed-width") %>  " +
+         "<%= j link_to("Edit", edit_issue_state_path(@issue_state), remote: true, class: "btn btn-primary btn-sm text-white mr-2 mb-1 mb-md-0 fixed-width") %>  " +
+         "<%= j link_to("Show", issue_state_path(@issue_state), remote: true, class: "btn btn-success btn-sm text-white fixed-width") %>"
+}
+
+function issueStateUpdateClassAndId(row) {
+  row.attr('id', "issue_state_<%= @issue_state.id %>");
+  row.attr('class', "<%= @issue_state.issue_id.nil? ? 'global' : 'issue-'+@issue_state.issue_id.to_s %>");
```
Comment (abubakarazeem) [inline app/views/issue_states/update.js.erb]: helper method

File: app/views/layouts/_errors.html.erb [hunk 39]
```diff
@@ -0,0 +1,9 @@
+
+<% model.errors.full_messages.each do |msg| %>
```
Comment (abubakarazeem) [inline app/views/layouts/_errors.html.erb]: msg --> error_message

File: app/controllers/issue_states_controller.rb (hunk 5 above)
Comment (Anmol79) [inline app/controllers/issue_states_controller.rb]: remove issue id from issue_state, discuss with me

**Pull request merged**